### ISSUE TITLE
Add schema validation and sanity checks to models

### ISSUE THREAD

**Issue by r-marques** [bigchaindb/bigchaindb#720]
_From @sohkai on September 30, 2016 11:38_

As of now, we do very little to ensure that the given inputs are what we expect them to be when constructing the models. We **really** should be doing at least some checks with json schema or any other validation approach; not doing so results in the **worst** and most **nefarious** types of bugs for users.

Earlier, I created a transfer transaction and accidentally passed through a keypair dict as part of the `owners_after` argument in the [driver](https://github.com/sbellem/bigchaindb-driver/) (rather than just a string verifying key). The driver was able to create, sign, and send this wrongly-formatted condition all the way to BigchainDB, who accepted it and wrote it onto RethinkDB (see below). Now my public/private keypair is forever visible to the world, and this condition is probably impossible to fulfill.

If I was a naive user and assumed everything had worked correctly, since nothing had told me otherwise, I might now start creating `n` transactions this way without ever realizing that I've completely shot myself in the foot. Oops.

What really should've happened is me getting an error immediately upon the `owners_after` signature being wrong. Crisis averted.

The persisted transaction, available in a block in RethinkDB:

```
{'id': '960a684bb92de50384ebf568b7e1e78d88f8afbb674872b3d25356eae77fba51',
 'transaction': {'conditions': [{'cid': 0,
                                 'condition': {'uri': {'signing_key': '4S1dzx3PSdMAfs59aBkQefPASizTs728HnhLNpYZWCad',
                                                       'verifying_key': '2dBVUoATxEzEqRdsi64AFsJnn2ywLCwnbNwW7K9BuVuS'}},
                                 'owners_after': [{'signing_key': '4S1dzx3PSdMAfs59aBkQefPASizTs728HnhLNpYZWCad',
                                                   'verifying_key': '2dBVUoATxEzEqRdsi64AFsJnn2ywLCwnbNwW7K9BuVuS'}]}],
                 'data': None,
                 'fulfillments': [{'fid': 0,
                                   'fulfillment': 'cf:4:4HwjqBgNkDK0fD1ajmFn0OZ75N3Jk-xIV2zlhgPxP2a7KeSZy56l5uU_Wqj6FmQz9O8xUYS3eIdR0RGDAobM4grfbtmeWbv4FXwtd5FyYbh6Azs4sle_eYFOKwsGREcL',
                                   'input': {'cid': 0,
                                             'txid': 'fad5ec32672b75cccf92992cc977f456bd587a7d49956fed67a227fb26d13250'},
                                   'owners_before': ['G7J7bXF8cqSrjrxUKwcF8tCriEKC5CgyPHmtGwUi4BK3']}],
                 'operation': 'TRANSFER',
                 'timestamp': '1475228620'},
 'version': 1}
```

**Notice the condition, which lists the entire given dict**:

```
{
    ...
    'conditions': [{'cid': 0,
                    'condition': {'uri': {'signing_key': '4S1dzx3PSdMAfs59aBkQefPASizTs728HnhLNpYZWCad',
                                          'verifying_key': '2dBVUoATxEzEqRdsi64AFsJnn2ywLCwnbNwW7K9BuVuS'}},
                    'owners_after': [{'signing_key': '4S1dzx3PSdMAfs59aBkQefPASizTs728HnhLNpYZWCad',
                                      'verifying_key': '2dBVUoATxEzEqRdsi64AFsJnn2ywLCwnbNwW7K9BuVuS'}]}],
    ...
```

(In fact, I probably could've put anything there; it just so happened that it was the keypair.)

---

At the very least, as suggested by @sbellem, it should be made very clear (with big warning signs and all that) that whatever you give that is turned into the `owners_after` is forever written in the transaction's condition. We could also add some sort of identity verification system, but that's out of scope.

cc @TimDaub @diminator 

_Copied from original issue: bigchaindb/bigchaindb-common#31_

**Comment by r-marques** [comment]
_From @diminator on September 30, 2016 13:45_

oh man, that would be terrible.

here are some options:
- only store the URI of the condition, together with `owners_after`. Exchange of details if off-chain
- do checks on custom conditions/`owners_after`

**Comment by r-marques** [comment]
_From @ttmc on October 3, 2016 9:21_

I agree that schema validation is important.

Here are some existing schema validation issues:

* Transaction schema validation - https://github.com/bigchaindb/bigchaindb/issues/275
* Block schema validation - https://github.com/bigchaindb/bigchaindb/issues/446
* Vote schema validation - https://github.com/bigchaindb/bigchaindb/issues/447

Back in May, @vrde did a first cut of transaction schema validation with https://github.com/bigchaindb/bigchaindb/pull/311 . As of this comment, that PR was not merged and was closed.

**Comment by r-marques** [comment]
_From @TimDaub on October 4, 2016 12:52_

Sanity checks should be done in the classes here.

With schema validation, I'm not sure where it should go.
As it's JSON schema validation, I'd say it would naturally go into driver and node. As we don't want to duplicate code it could hence go in BDBC.

Now, I think still that this is a multi layered problem.

1. BDB-node: Does is mark this transaction as valid (if so, that's a bug and needs to be fixed in the node. Would be nice if you could check and create a ticket in the case @sohkai)

2. BDB-C: Do we "need" `owners_afters`, `owners_before`: Kind of. The BDB node relies heavily on querying for public keys (which might be the problem to begin with). This is due to the fact that Interledger's cryptoconditions do not support by-owner-indexable transactions.
This is I think the biggest part of the problem, which we should try to solve.

Possible solutions:

- Only have URIs for a fulfillment and condition
- On every serialization sanity check against the `owners_*`: https://github.com/bigchaindb/bigchaindb-common/issues/15
- Modify the Cyptocondition URIs so much that we can add the `owners_*` as an optional parameter: think of a URLs query parameters.
- Create a off-chain mechanism for `receiving` transactions (which I think doesn't fit at all with BDB's philosophy and is kind of a weak technical answer as this is simply _not how blockchains work_ (haters gonna hate).
- ... (more creative input?)

**Comment by r-marques** [comment]
_From @sohkai on October 5, 2016 12:14_

@diminator:
>
- only store the URI of the condition, together with owners_after. Exchange of details if off-chain
- do checks on custom conditions/owners_after

Not sure about all the details surrounding this, but it seems like `owners_after` is quite a messy topic. As for "custom conditions," I'm not sure what this might mean; I suspect this would make more sense if we provided prebuilt conditions?

@TimDaub:
> BDB-node: Does is mark this transaction as valid

I'm 99% sure it's marking it as valid, since the transaction I showed was in the 'bigchain' table and not the backlog table. BigchainDB also didn't complain about validation.

>
- Only have URIs for a fulfillment and condition
- On every serialization sanity check against the owners_*

These would be super cool to have.

> With schema validation, I'm not sure where it should go.

If we assume that BigchainDB shouldn't worry about transactions and their implementation, I think it makes sense to validate the transactions here. The voters and blocks can be validated in the server code.

----------

A good place to bring this up is when we discuss https://github.com/bigchaindb/bigchaindb-common/issues/12.
